ENH: Support kurtosis (kurt) in DataFrameGroupBy and SeriesGroupBy

snitish opened the pull request:
- [x] closes #40139 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

DataFrameGroupBy and SeriesGroupBy currently support mean, std and skew (the first 3 moments) but not kurtosis (the 4th moment). This change addresses that. I implemented kurtosis in cython in similar fashion to skewness. I've verified that the output of this function matches that of DataFrame.kurt().

snitish committed:
ENH: Support kurtosis (kurt) in DataFrameGroupBy and SeriesGroupBy

rhshadrach commented on pandas/core/groupby/generic.py:
```diff
@@ -1279,6 +1279,84 @@ def alt(obj):
+        Name: Max Speed, dtype: float64
+        """
+
+        def alt(obj):
+            # This should not be reached since the cython path should raise
+            #  TypeError and not NotImplementedError.
+            raise TypeError(f"'kurt' is not supported for dtype={obj.dtype}")
```
I see that this is similar to what's done for skew, but we can also now provide `alt=None` instead of providing an unnecessary fallback; I think this should be done for both of these ops. Okay to do it for skew here too if you'd like.

snitish replied:
Updated with `alt=None` for both kurt and skew.

rhshadrach commented on pandas/tests/groupby/methods/test_kurt.py:
```diff
@@ -0,0 +1,27 @@
+
+    grpwise = [grp.kurt().to_frame(i).T for i, grp in gb]
+    expected = pd.concat(grpwise, axis=0)
+    expected.index = expected.index.astype(result.index.dtype)  # 32bit builds
```
Can you use `"intp"` here instead of having the expected depend on the result.

snitish replied:
Fixed

snitish replied:
The test seems to fail in some environments after I changed it to `"intp"`. Do I need to separately assert the index and the values like some other tests do?

rhshadrach replied:
Sorry, should be int64 I think.

rhshadrach commented on pandas/tests/groupby/methods/test_kurt.py:
```diff
@@ -0,0 +1,27 @@
+
+
+def test_groupby_kurt_equivalence():
+    # Test that that groupby kurt method (which uses libgroupby.group_kurt)
```
Can you start tests with a comment referencing the GitHub issue.

    # GH#40139

snitish replied:
Added issue number.

rhshadrach reviewed (changes requested):
Very nice! A few test requests (I think these are not covered yet):

- skipna with NA values in the data
- Float64 and float64[pyarrow] dtypes
- Constant data (e.g. `[1, 1, 1, 1]`)

snitish committed:
ENH: Address review comments

snitish commented on pandas/_libs/groupby.pyx:
```diff
@@ -995,6 +995,100 @@ def group_skew(
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.cdivision(True)
+@cython.cpow
```
Fixed it for both kurt and skew.

snitish commented on pandas/_libs/groupby.pyx:
```diff
@@ -995,6 +995,100 @@ def group_skew(
+
+                if not isna_entry:
+                    # Based on RunningStats::Push from
+                    #  https://www.johndcook.com/blog/skewness_kurtosis/
```
Updated comment with a note explaining that only the running update is based on that blog.

snitish commented:
Thanks for the review @rhshadrach. 

- Addressed your comments
- Added test case for skipna=False (by default it's true)
- Added test case for float64[pyarrow] (we already have one for float64)
- Added test case for constant data. Note that the result here is 0.0, consistent with `DataFrame.kurt()` and `Series.kurt()`

snitish committed:
ENH: Fix comments in new test cases

snitish committed:
ENH: Skip pyarrow test case if no pyarrow available

snitish committed:
ENH: Update to intp instead of np.intp

snitish committed:
ENH: Change intp to int64

snitish committed:
Merge branch 'main' into kurtosis

rhshadrach reviewed (changes requested):
> Added test case for float64[pyarrow] (we already have one for float64)

The request was for `Float64`, the NumPy-nullable array. Can just parameterize your arrow test I think:

```python
@pytest.mark.parametrize("dtype", [pytest.param("float64[pyarrow]", marks=td.skip_if_no("pyarrow")), "Float64")
```

snitish committed:
Address review comments

rhshadrach commented:
Apologies @snitish - this fell off my radar. Should be able to get to it today or tomorrow.

rhshadrach reviewed (approved):
lgtm - will merge in two days unless there are further reviews.

snitish commented:
Thank you @rhshadrach!

mroeschke commented:
Thanks @snitish
